# Predicting NBA Game Winners

In [1]:
# Load necessary libraries

import platform
import numpy as np
import pandas as pd
import sklearn as sk
import os

In [2]:
# Load the dataset: NBA statistics from the 2013-2014 basketball season
df = pd.read_csv('./datasets/NBA Regular Season Results 2013-2014.csv')

In [3]:
from sklearn.metrics import f1_score, make_scorer, classification_report

scorer = make_scorer(f1_score, pos_label = None, average = 'weighted')

In [4]:
# Examine the dataset with its initial formatting
df.head()

Date Start (ET)       Visitor/Neutral  PTS         Home/Neutral  \
0  Tue Oct 29 2013    7:00 pm         Orlando Magic   87       Indiana Pacers   
1  Tue Oct 29 2013   10:30 pm  Los Angeles Clippers  103   Los Angeles Lakers   
2  Tue Oct 29 2013    8:00 pm         Chicago Bulls   95           Miami Heat   
3  Wed Oct 30 2013    7:00 pm         Brooklyn Nets   94  Cleveland Cavaliers   
4  Wed Oct 30 2013    8:30 pm         Atlanta Hawks  109     Dallas Mavericks   

   PTS.1 Unnamed: 6 Unnamed: 7  Attend. Notes  
0     97  Box Score        NaN    18165   NaN  
1    116  Box Score        NaN    18997   NaN  
2    107  Box Score        NaN    19964   NaN  
3     98  Box Score        NaN    20562   NaN  
4    118  Box Score        NaN    19834   NaN

In [5]:
df.dtypes

# Note the Date is read as an object field

Date               object
Start (ET)         object
Visitor/Neutral    object
PTS                 int64
Home/Neutral       object
PTS.1               int64
Unnamed: 6         object
Unnamed: 7         object
Attend.             int64
Notes              object
dtype: object

In [6]:
# Don't read the first row because it is blank, and read the date column as a date

df = pd.read_csv('./datasets/NBA Regular Season Results 2013-2014.csv', parse_dates = ['Date'])

# Correct column names
df.columns = ['Date', 'Start Time (ET)', 'Visitor Team', 'Visitor Score', 'Home Team', 'Home Score', 'Box Score', 
              'OT', 'Attendance', 'Notes']

In [7]:
df.shape

(1230, 10)

In [8]:
# View the updated dataset
df.head()

Date Start Time (ET)          Visitor Team  Visitor Score  \
0 2013-10-29         7:00 pm         Orlando Magic             87   
1 2013-10-29        10:30 pm  Los Angeles Clippers            103   
2 2013-10-29         8:00 pm         Chicago Bulls             95   
3 2013-10-30         7:00 pm         Brooklyn Nets             94   
4 2013-10-30         8:30 pm         Atlanta Hawks            109   

             Home Team  Home Score  Box Score   OT  Attendance Notes  
0       Indiana Pacers          97  Box Score  NaN       18165   NaN  
1   Los Angeles Lakers         116  Box Score  NaN       18997   NaN  
2           Miami Heat         107  Box Score  NaN       19964   NaN  
3  Cleveland Cavaliers          98  Box Score  NaN       20562   NaN  
4     Dallas Mavericks         118  Box Score  NaN       19834   NaN

In [9]:
# The Date field has been updated to read as a datetime field
# May not do anything with the field yet
df.dtypes

Date               datetime64[ns]
Start Time (ET)            object
Visitor Team               object
Visitor Score               int64
Home Team                  object
Home Score                  int64
Box Score                  object
OT                         object
Attendance                  int64
Notes                      object
dtype: object

In [10]:
# Create new feature that notes whether the home team won
df['Home Win'] = df['Visitor Score'] < df['Home Score']

# Assign "class values" to the Home Win column

### This will also be the value that we want to predict ###
y_true = df['Home Win'].values

In [11]:
# Review new column with correctly populated fields
df.head()

Date Start Time (ET)          Visitor Team  Visitor Score  \
0 2013-10-29         7:00 pm         Orlando Magic             87   
1 2013-10-29        10:30 pm  Los Angeles Clippers            103   
2 2013-10-29         8:00 pm         Chicago Bulls             95   
3 2013-10-30         7:00 pm         Brooklyn Nets             94   
4 2013-10-30         8:30 pm         Atlanta Hawks            109   

             Home Team  Home Score  Box Score   OT  Attendance Notes  Home Win  
0       Indiana Pacers          97  Box Score  NaN       18165   NaN      True  
1   Los Angeles Lakers         116  Box Score  NaN       18997   NaN      True  
2           Miami Heat         107  Box Score  NaN       19964   NaN      True  
3  Cleveland Cavaliers          98  Box Score  NaN       20562   NaN      True  
4     Dallas Mavericks         118  Box Score  NaN       19834   NaN      True

## What is the baseline?

How frequently do the home teams win? Typically, the assumption is that home teams will win more often than not

In [12]:
# Establishing the baseline win percentage for the Home Teams overall
n_games = df['Home Win'].count()
n_homewins = df['Home Win'].sum()
win_percentage = n_homewins / n_games

print('Home Win percentage: {0:.2f}%'.format(100 * win_percentage))

Home Win percentage: 58.05%


The home teams' win percentage is 58.0%. If a model is considered useful, it must (at minimum) have a better score than the established baseline

In [13]:
# Predicting the baseline for Home Teams wins with simple classifier

from sklearn.metrics import f1_score

y_pred = [1] * len(y_true)
print('F1: {0:.4f}%'.format(f1_score(y_true, y_pred, pos_label = None, average = 'weighted') * 100))

F1: 42.6408%


/Users/slic_vic/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


The calculated F1 score of 42.6% is the score to beat.

## Creating More Features

In [14]:
# Create two new columns, with all of the values set to False
df['Home Last Win'] = False
df['Visitor Last Win'] = False

In [15]:
# Determining whether the home and visitor teams won their last games
# This will update the Home Last Win & Visitor Last Win columns to either remain False or change to True

from collections import defaultdict

won_last = defaultdict(int)   # The default value of int is 0

# Inefficient method to deterime previous wins
for index, row in df.iterrows():
    home_team = row['Home Team']
    visitor_team = row['Visitor Team']
    row['Home Last Win'] = won_last[home_team]
    row['Visitor Last Win'] = won_last[visitor_team]
    # Set current win
    won_last[home_team] = row['Home Win']
    won_last[visitor_team] = not row['Home Win']
df.loc[20:25]

Date Start Time (ET)            Visitor Team  Visitor Score  \
20 2013-11-01         7:30 pm         Milwaukee Bucks            105   
21 2013-11-01         8:00 pm              Miami Heat            100   
22 2013-11-01         7:00 pm     Cleveland Cavaliers             84   
23 2013-11-01         9:00 pm  Portland Trail Blazers            113   
24 2013-11-01         8:00 pm        Dallas Mavericks            105   
25 2013-11-01        10:30 pm       San Antonio Spurs             91   

             Home Team  Home Score  Box Score   OT  Attendance Notes  \
20      Boston Celtics          98  Box Score  NaN       18624   NaN   
21       Brooklyn Nets         101  Box Score  NaN       17732   NaN   
22   Charlotte Bobcats          90  Box Score  NaN       18017   NaN   
23      Denver Nuggets          98  Box Score  NaN       19155   NaN   
24     Houston Rockets         113  Box Score  NaN       18142   NaN   
25  Los Angeles Lakers          85  Box Score  NaN       18997   NaN   

    Home Win  Home Last Win  Visitor Last Win  
20     False          False             False  
21      True          False             False  
22      True          False             False  
23     False          False             False  
24      True          False             False  
25     False          False             False

In [16]:
df.loc[90:95][['Date', 'Visitor Team', 'Visitor Score', 'Home Team', 'Home Score', 
               'Home Win', 'Home Last Win', 'Visitor Last Win']]

Date            Visitor Team  Visitor Score              Home Team  \
90 2013-11-09        Dallas Mavericks             91        Milwaukee Bucks   
91 2013-11-09  Portland Trail Blazers             96       Sacramento Kings   
92 2013-11-09               Utah Jazz             91        Toronto Raptors   
93 2013-11-10  Minnesota Timberwolves            113     Los Angeles Lakers   
94 2013-11-10       San Antonio Spurs            120        New York Knicks   
95 2013-11-10      Washington Wizards            105  Oklahoma City Thunder   

    Home Score  Home Win  Home Last Win  Visitor Last Win  
90          83     False          False             False  
91          85     False          False             False  
92         115      True          False             False  
93          90     False          False             False  
94          89     False          False             False  
95         106      True          False             False

# Basic Classification with a Decision Tree

The default settings for the Decision Tree will be used for the home wins, winning streaks, and rankings

In [17]:
# Basic Decision Tree Classifier set up
from sklearn.tree import DecisionTreeClassifier
dtc = DecisionTreeClassifier(random_state = 14) # Remove random_state to get non-replicable results

In [18]:
from sklearn.cross_validation import cross_val_score

# Use selected features as input for the classifier (target)
X_previouswins = df[['Home Last Win', 'Visitor Last Win']].values

# Decision Tree Classifier
dtc = DecisionTreeClassifier(random_state = 14)
scores = cross_val_score(dtc, X_previouswins, y_true, scoring = scorer)

# Print results
print("Using just the last result from the home and visitor teams")
print('F1: {0:.4f}%'.format(np.mean(scores) * 100))

Using just the last result from the home and visitor teams
F1: 42.6408%


/Users/slic_vic/anaconda3/lib/python3.6/site-packages/sklearn/cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)
/Users/slic_vic/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
/Users/slic_vic/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
/Users/slic_vic/anaconda3/lib/python3.6/site-packages/sklearn/met

When using the features that deterime whether a team won their previous game, the result is 42.6%, which is equal to the baseline score

There was no improvement.


*** In the lecture, the result here was closer to 50-51%. What is happening here for the Decision Tree to result in the same score as the baseline? ***

### Decision Tree Classifier: Considerating Winning Streaks

In [19]:
# Taking into consideration winning streaks - What are the teams' win streaks coming into the game?
df["Home Win Streak"] = 0
df["Visitor Win Streak"] = 0

# Did the home and visitor teams win their last game?
from collections import defaultdict
win_streak = defaultdict(int)

for index, row in df.iterrows():  # Note that this is not the most efficient method
    home_team = row["Home Team"]
    visitor_team = row["Visitor Team"]
    row["Home Win Streak"] = win_streak[home_team]
    row["Visitor Win Streak"] = win_streak[visitor_team]
    df.loc[index] = row    
    
    # Set current win streak
    if row["Home Win"]:
        win_streak[home_team] += 1
        win_streak[visitor_team] = 0
    else:
        win_streak[home_team] = 0
        win_streak[visitor_team] += 1

In [20]:
df.loc[50:60][['Date', 'Visitor Team', 'Visitor Score', 'Home Team', 'Home Score', 
               'Home Win', 'Home Win Streak', 'Visitor Win Streak']]

Date        Visitor Team  Visitor Score               Home Team  \
50 2013-11-05           Utah Jazz             88           Brooklyn Nets   
51 2013-11-05  Los Angeles Lakers            104        Dallas Mavericks   
52 2013-11-05   San Antonio Spurs            102          Denver Nuggets   
53 2013-11-05      Indiana Pacers             99         Detroit Pistons   
54 2013-11-05        Phoenix Suns            104    New Orleans Pelicans   
55 2013-11-05   Charlotte Bobcats            102         New York Knicks   
56 2013-11-05     Houston Rockets            116  Portland Trail Blazers   
57 2013-11-05       Atlanta Hawks            105        Sacramento Kings   
58 2013-11-05          Miami Heat            104         Toronto Raptors   
59 2013-11-06           Utah Jazz             87          Boston Celtics   
60 2013-11-06     Toronto Raptors             90       Charlotte Bobcats   

    Home Score  Home Win  Home Win Streak  Visitor Win Streak  
50         104      True                0                   0  
51         123      True                1                   1  
52          94     False                0                   0  
53          91     False                1                   3  
54          98     False                1                   0  
55          97     False                0                   0  
56         101     False                2                   0  
57         100     False                0                   0  
58          95     False                1                   1  
59          97      True                0                   0  
60          92      True                1                   0

In [21]:
# Use selected features as input for the classifier (target)
X_winstreak = df[["Home Last Win", "Visitor Last Win", "Home Win Streak", "Visitor Win Streak"]].values

# Decision Tree Classifier
dtc = DecisionTreeClassifier(random_state = 14)
scores = cross_val_score(dtc, X_winstreak, y_true, scoring = scorer)

# Print results
print("Using whether the home team is ranked higher")
print("F1: {0:.4f}%".format(np.mean(scores) * 100))

Using whether the home team is ranked higher
F1: 52.4982%


When taking into consideration whether a team is on a win streak, the model's results show 52.5%

### Decision Tree Classifier: Based on Previous Season Standings

In [22]:
# Identify which team is higher in the standings, based on the previous year's regular season final standings
# Load the standings data file

rank = pd.read_csv('./datasets/2012-2013 Regular Season Standings.csv', index_col = 'Team')

In [23]:
rank.head(3)

Rk Overall  Home   Road      E      W     A     C  \
Team                                                                       
Miami Heat              1   66-16  37-4  29-12  41-11   25-5  14-4  12-6   
Oklahoma City Thunder   2   60-22  34-7  26-15   21-9  39-13   7-3   8-2   
San Antonio Spurs       3   58-24  35-6  23-18   25-5  33-19   8-2   9-1   

                         SE    NW ...    Post   ≤3    ≥10  Oct   Nov   Dec  \
Team                              ...                                        
Miami Heat             15-1   8-2 ...    30-2  9-3   39-8  1-0  10-3  10-5   
Oklahoma City Thunder   6-4  10-6 ...    21-8  3-6   44-6  NaN  13-4  11-2   
San Antonio Spurs       8-2   9-9 ...   16-12  9-5  31-10  1-0  12-4  12-4   

                        Jan   Feb   Mar  Apr  
Team                                          
Miami Heat              8-5  12-1  17-1  8-1  
Oklahoma City Thunder  11-5   7-4  12-5  6-2  
San Antonio Spurs      12-3   8-3  10-4  3-6  

[3 rows x 23 columns]

In [24]:
# Rename columns in the rank DataFrame
rank.columns = ['Rank', 'Overall', 'Home', 'Road', 'Eastern Conference', 'Western Conference', 
                'Atlantic Div', 'Central Div', 'Southeast Div', 'Northwest Div', 'Pacific Div', 'Southwest Div', 
                'Pre All-Star', 'Post All-Star', 'Margin ≤3', 'Margin ≥10', 'Oct', 'Nov', 'Dec', 'Jan', 'Feb', 
                'Mar', 'Apr']

In [25]:
rank.head(3)

Rank Overall  Home   Road Eastern Conference  \
Team                                                                  
Miami Heat                1   66-16  37-4  29-12              41-11   
Oklahoma City Thunder     2   60-22  34-7  26-15               21-9   
San Antonio Spurs         3   58-24  35-6  23-18               25-5   

                      Western Conference Atlantic Div Central Div  \
Team                                                                
Miami Heat                          25-5         14-4        12-6   
Oklahoma City Thunder              39-13          7-3         8-2   
San Antonio Spurs                  33-19          8-2         9-1   

                      Southeast Div Northwest Div ...  Post All-Star  \
Team                                              ...                  
Miami Heat                     15-1           8-2 ...           30-2   
Oklahoma City Thunder           6-4          10-6 ...           21-8   
San Antonio Spurs               8-2           9-9 ...          16-12   

                      Margin ≤3 Margin ≥10  Oct   Nov   Dec   Jan   Feb   Mar  \
Team                                                                            
Miami Heat                  9-3       39-8  1-0  10-3  10-5   8-5  12-1  17-1   
Oklahoma City Thunder       3-6       44-6  NaN  13-4  11-2  11-5   7-4  12-5   
San Antonio Spurs           9-5      31-10  1-0  12-4  12-4  12-3   8-3  10-4   

                       Apr  
Team                        
Miami Heat             8-1  
Oklahoma City Thunder  6-2  
San Antonio Spurs      3-6  

[3 rows x 23 columns]

In [26]:
# Create a new feature -> Home Team Ranks Higher
# First create a function that iterates through the df to determine if the home team has a higher rank based on rank df

def home_team_ranks_higher(row):
    home_team = row["Home Team"]
    visitor_team = row["Visitor Team"]
    
    # Adjusting the New Orleans team names due to off-season league changes between 12-13 & 13-14
    if home_team == "New Orleans Pelicans":
        home_team = "New Orleans Hornets"
    if visitor_team == "New Orleans Pelicans":
        visitor_team = "New Orleans Hornets"
    
    home_rank = rank.loc[home_team]["Rank"]
    visitor_rank = rank.loc[visitor_team]["Rank"]
    
    return home_rank < visitor_rank   # The higher ranking will be the lower number

In [27]:
# Create new column by using .apply(home_team_ranks_higher) across the DataFrame
df["Home Team Ranks Higher"] = df.apply(home_team_ranks_higher, axis = 1)

# Display the selected columns on the sliced DataFrame (first 5 rows)
df[:5][['Date', 'Visitor Team', 'Visitor Score', 'Home Team', 'Home Score', 'Home Win', 'Home Team Ranks Higher']]

Date          Visitor Team  Visitor Score            Home Team  \
0 2013-10-29         Orlando Magic             87       Indiana Pacers   
1 2013-10-29  Los Angeles Clippers            103   Los Angeles Lakers   
2 2013-10-29         Chicago Bulls             95           Miami Heat   
3 2013-10-30         Brooklyn Nets             94  Cleveland Cavaliers   
4 2013-10-30         Atlanta Hawks            109     Dallas Mavericks   

   Home Score  Home Win  Home Team Ranks Higher  
0          97      True                    True  
1         116      True                   False  
2         107      True                    True  
3          98      True                   False  
4         118      True                   False

______________________________________________________________________________________________________________________

In [28]:
# # Add new column without using a function & .apply
# df["Home Team Ranks Higher"] = 0

# for index, row in df.iterrows():
#     home_team = row["Home Team"]
#     visitor_team = row["Visitor Team"]
   
#     # Adjusting the New Orleans team names due to off-season league changes between 12-13 & 13-14
#     if home_team == "New Orleans Pelicans":
#         home_team = "New Orleans Hornets"
#     elif visitor_team == "New Orleans Pelicans":
#         visitor_team = "New Orleans Hornets"
    
#     home_rank = rank[rank["Team"] == home_team]["Rank"].values[0]
#     visitor_rank = rank[rank["Team"] == visitor_team]["Rank"].values[0]
#     row["Home Team Ranks Higher"] = int(home_rank > visitor_rank)
#     df.loc[index] = row

In [29]:
# df[:5][['Date', 'Visitor Team', 'Visitor Score', 'Home Team', 'Home Score', 'Home Win']]

______________________________________________________________________________________________________________________

In [30]:
# Decision Tree Classifier based on if Home Team has a Higher Ranking

# Use selected features as input for the classifier (target)
X_homehigher = df[["Home Last Win", "Visitor Last Win", "Home Team Ranks Higher"]].values

# Decision Tree Classifier
dtc = DecisionTreeClassifier(random_state = 14)
scores = cross_val_score(dtc, X_homehigher, y_true, scoring = scorer)

# Print results
print("Using whether the home team is ranked higher")
print("F1: {0:.4f}%".format(np.mean(scores) * 100))

Using whether the home team is ranked higher
F1: 59.4525%


When taking into consideration whether the home team has a higher rank in the previous season's standings, the model's results show 59.45%


*** This result is .2 higher than the lecture ***

In [31]:
# Adjusting the parameters of the algorithm using GridSearchCV to test if there is an improvement in the model's score

from sklearn.grid_search import GridSearchCV

parameter_space = {
    'max_depth': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20],
}
dtc = DecisionTreeClassifier(random_state = 14)

grid = GridSearchCV(dtc, parameter_space, scoring = scorer)
grid.fit(X_homehigher, y_true)

print('F1: {0:.4f}%'.format(grid.best_score_ * 100))

F1: 59.4525%


/Users/slic_vic/anaconda3/lib/python3.6/site-packages/sklearn/grid_search.py:42: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. This module will be removed in 0.20.
  DeprecationWarning)


By adjusting the parameters (tuning) of the Decision Tree, there is no change in the models

## Decision Tree Classifier: Based on whether the Home Team Won the Last Matchup

______________________________________________________________________________________________________________________

*** This function was entered exactly, but when applied, there is an error message. Used the inefficient method to continue with the lecture/notebook ***

*** Skip to 22:20 in video ***

In [32]:
# Build a function that determines whether a team won the last matchup between the 2 teams
# This does not take into consideration the home/visitor teams

last_game_winner = defaultdict(int)

def home_team_won_last(row):
    home_team = row['Home Team']
    visitor_team = row['Visistor Team']
    
    # Sort for a consistent ordering
    teams = tuple(sorted([home_team, visitor_team]))
    # Parse the row for which team won the last matchup, then add a 1 if the Home Team won
    result = 1 if last_game_winner[teams] == row['Home Team'] else 0
    
    # Update record for next matchup
    winner = row['Home Team'] if row['Home Win'] else row['Visitor Team']
    last_game_winner[teams] = winner
    
    return result

# Function works, but is not applying across DataFrame (see cell below)

In [33]:
# Create new column by using .apply(home_team_won_last) across the DataFrame
# df['Home Team Won Last'] = df.apply(home_team_won_last, axis = 1)

In [34]:
# Display the selected columns on the sliced DataFrame (5 random rows)
# df[90:100][['Date', 'Visitor Team', 'Visitor Score', 'Home Team', 'Home Score', 'Home Win', 'Home Team Won Last']]

______________________________________________________________________________________________________________________

In [35]:
# Add a new feature without using function -> Determines whether a team won the last matchup between the 2 teams
# This does not take into consideration the home/visitor teams

last_game_winner = defaultdict(int)
df["Home Team Won Last"] = 0

for index, row in df.iterrows():
    home_team = row["Home Team"]
    visitor_team = row["Visitor Team"]
    teams = tuple(sorted([home_team, visitor_team]))  # Sort for a consistent ordering
    
    # Set in the row, who won the last encounter
    row["Home Team Won Last"] = 1 if last_game_winner[teams] == row["Home Team"] else 0
    df.loc[index] = row
    
    # Who won this one?
    winner = row["Home Team"] if row["Home Win"] else row["Visitor Team"]
    last_game_winner[teams] = winner

In [36]:
# Display the selected columns on the sliced DataFrame (10 random rows)
df[90:100][['Date', 'Visitor Team', 'Visitor Score', 'Home Team', 'Home Score', 'Home Win', 'Home Team Won Last']]

Date            Visitor Team  Visitor Score              Home Team  \
90 2013-11-09        Dallas Mavericks             91        Milwaukee Bucks   
91 2013-11-09  Portland Trail Blazers             96       Sacramento Kings   
92 2013-11-09               Utah Jazz             91        Toronto Raptors   
93 2013-11-10  Minnesota Timberwolves            113     Los Angeles Lakers   
94 2013-11-10       San Antonio Spurs            120        New York Knicks   
95 2013-11-10      Washington Wizards            105  Oklahoma City Thunder   
96 2013-11-10    New Orleans Pelicans             94           Phoenix Suns   
97 2013-11-11           Orlando Magic            105         Boston Celtics   
98 2013-11-11           Atlanta Hawks            103      Charlotte Bobcats   
99 2013-11-11     Cleveland Cavaliers             81          Chicago Bulls   

    Home Score  Home Win  Home Team Won Last  
90          83     False                   0  
91          85     False                   0  
92         115      True                   0  
93          90     False                   0  
94          89     False                   0  
95         106      True                   0  
96         101      True                   1  
97         120      True                   1  
98          94     False                   0  
99          96      True                   0

In [37]:
# Use selected features as input for the classifier (target)
X_home_higher = df[['Home Last Win', 'Visitor Last Win', "Home Team Ranks Higher", "Home Team Won Last"]].values

# Decision Tree Classifier
dtc = DecisionTreeClassifier(random_state = 14)
scores = cross_val_score(dtc, X_home_higher, y_true, scoring = scorer)

# Print results
print("Using whether the home team won the last matchup")
print("F1: {0:.4f}%".format(np.mean(scores) * 100))

Using whether the home team won the last matchup
F1: 60.1726%


After adding the feature that takes into account whether the home team won the last matchup between the teams, the model's accuracy increased to 60.17%

*** This is about .09 lower than the lecture ***

### Decision Tree Classifier: Encoding Teams as Features

In [38]:
df['Home Team'].values

array(['Indiana Pacers', 'Los Angeles Lakers', 'Miami Heat', ...,
       'Portland Trail Blazers', 'Sacramento Kings', 'San Antonio Spurs'],
      dtype=object)

In [39]:
# Use LabelEncoder to transform the teams by assigning numbers to them
from sklearn.preprocessing import LabelEncoder, OneHotEncoder

encoding = LabelEncoder()
encoding.fit(df["Home Team"].values)

LabelEncoder()

In [40]:
# Numerical values assigned to the home and visitor teams
home_teams = encoding.transform(df["Home Team"].values)
visitor_teams = encoding.transform(df["Visitor Team"].values)

# Stacks arrays in a sequence vertically (row-wise) & transposes the data
X_teams = np.vstack([home_teams, visitor_teams]).T

In [41]:
# OneHotEncoder takes each of the features of the numerical valued-teams & identify if it was the home or visitor team
# Will display a 1 if True, 0 if False
onehot = OneHotEncoder()
X_teams = onehot.fit_transform(X_teams).todense()

In [42]:
# View the data size of X_teams
X_teams.shape

(1230, 60)

In [43]:
# Print the first 30 elements in X_teams
print('Home:', X_teams[0:30])
print('Visitor:', X_teams[0,30:])

Home: [[0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 1. 0.]]
Visitor: [[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0.
  0. 0. 0. 0. 0. 0.]]


In [44]:
# Run the classifier on the new teams as features
# Decision Tree Classifier
dtc = DecisionTreeClassifier(random_state = 14)
scores = cross_val_score(dtc, X_teams, y_true, scoring = scorer)

# Print results
print("F1: {0:.4f}%".format(np.mean(scores) * 100))

F1: 59.9439%


When encoding the teams as individual features, the accuracy of the model resulted in an F1 score of 59.94%.

*** This result is .32 higher than the lecture ***

## Random Forest Classifiers

### Random Forest Classifier: Use ONLY the Teams Encoded as Features

In [45]:
# Import Random Forest Classifier
from sklearn.ensemble import RandomForestClassifier

# Random Forest Classifier
rfc = RandomForestClassifier(random_state = 14)
scores = cross_val_score(rfc, X_teams, y_true, scoring = scorer)

# Print results
print("Using full team labels is ranked higher")
print("F1: {0:.4f}%".format(np.mean(scores) * 100))

Using full team labels is ranked higher
F1: 60.8530%


*** This result is .4 higher than the lecture ***

### Random Forest Classifier: Adjusting Parameters with GridSearchCV for Encoded Teams

In [46]:
# Set up more detailed parameters for the Random Forest Classifier in an attempt to improve accuracy score
# Adjusting the parameters of the algorithm using GridSearchCV

parameter_space = {
                   "max_features": [2, 10, 50, 'auto'],
                   "n_estimators": [50, 100, 200],
                   "criterion": ["gini", "entropy"],
                   "min_samples_leaf": [1, 2, 4, 6],
                   }

# Random Forest Classifier
rfc = RandomForestClassifier(random_state = 14)
grid = GridSearchCV(rfc, parameter_space, scoring = scorer)
grid.fit(X_teams, y_true)

# Print results
print("F1: {0:.4f}%".format(grid.best_score_ * 100))
print(grid.best_estimator_)

F1: 64.3611%
RandomForestClassifier(bootstrap=True, class_weight=None, criterion='entropy',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=4, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=200, n_jobs=1,
            oob_score=False, random_state=14, verbose=0, warm_start=False)


After adjusting the parameters when using the encoded teams as features, the model's results improved to 64.36%

### Random Forest Classifier: Use ALL Features for Data

In [47]:
# Combine the team features (from one hot endoder) with the other previously determined features
X_all = np.hstack([X_home_higher, X_teams])
print(X_all.shape)

(1230, 64)


In [48]:
# Random Forest Classifier
rfc = RandomForestClassifier(random_state = 14)
scores = cross_val_score(rfc, X_all, y_true, scoring = scorer)

# Print results
print("F1: {0:.4f}%".format(np.mean(scores) * 100))

F1: 61.4918%


There was a slight improvement to 61.49% with the additional features used in addition to the team features

*** This result is 1.1 higher than the lecture results ***

In [49]:
# Set up more detailed parameters for the Random Forest Classifier in an attempt to improve accuracy score
# Adjusting the parameters of the algorithm using GridSearchCV

parameter_space = {
                   "max_features": [2, 10, 50, 'auto'],
                   "n_estimators": [50, 100, 200],
                   "criterion": ["gini", "entropy"],
                   "min_samples_leaf": [1, 2, 4, 6],
                   }

# Random Forest Classifier
rfc = RandomForestClassifier(random_state = 14)
grid = GridSearchCV(rfc, parameter_space, scoring = scorer)
grid.fit(X_all, y_true)

# Print results
print("F1: {0:.4f}%".format(grid.best_score_ * 100))
print(grid.best_estimator_)

F1: 62.9018%
RandomForestClassifier(bootstrap=True, class_weight=None, criterion='entropy',
            max_depth=None, max_features=10, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=2, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=200, n_jobs=1,
            oob_score=False, random_state=14, verbose=0, warm_start=False)


After modifying the parameters of the Random Forest Classifier and using all of the features, the result was 62.9%. While this wasn't the best performing RFC, it did perform better than the initial baseline and the majority of the Decision Tree Classifiers.

*** This result is 1 lower than the lecture results ***

*** INSERT GRAPHS COMPARING MODEL SCORE RESULTS ***

# Testing on 2014-2015 Season

In [50]:
# Load the data for the 2014-15 season
df_21 = pd.read_csv('./datasets/NBA Regular Season Results 2014-2015.csv', parse_dates = ['Date'], 
                    dtype = {'Attendance':int}) # Attendence is still showing up as a float (HOW TO FIX?)

# Correct column names
df_21.columns = ['Date', 'Start Time (ET)', 'Visitor Team', 'Visitor Score', 'Home Team', 'Home Score', 'Box Score', 
              'OT', 'Attendance', 'Notes']

df_21.head(7)

Date Start Time (ET)        Visitor Team  Visitor Score  \
0 2014-10-28        10:30 pm     Houston Rockets            108   
1 2014-10-28         8:00 pm       Orlando Magic             84   
2 2014-10-28         8:00 pm    Dallas Mavericks            100   
3 2014-10-29         7:30 pm       Brooklyn Nets            105   
4 2014-10-29         7:00 pm     Milwaukee Bucks            106   
5 2014-10-29         9:00 pm     Detroit Pistons             79   
6 2014-10-29         7:00 pm  Philadelphia 76ers             91   

              Home Team  Home Score  Box Score   OT  Attendance Notes  
0    Los Angeles Lakers          90  Box Score  NaN     18997.0   NaN  
1  New Orleans Pelicans         101  Box Score  NaN     17097.0   NaN  
2     San Antonio Spurs         101  Box Score  NaN     19615.0   NaN  
3        Boston Celtics         121  Box Score  NaN     18624.0   NaN  
4     Charlotte Hornets         108  Box Score   OT     19439.0   NaN  
5        Denver Nuggets          89  Box Score  NaN     17136.0   NaN  
6        Indiana Pacers         103  Box Score  NaN     18165.0   NaN

In [51]:
# Load the previous season's standings
rank_20 = pd.read_csv('./datasets/2013-2014 Regular Season Standings.csv', index_col = 'Team')

# Rename columns in the rank DataFrame
rank_20.columns = ['Rank', 'Overall', 'Home', 'Road', 'Eastern Conference', 'Western Conference', 
                'Atlantic Div', 'Central Div', 'Southeast Div', 'Northwest Div', 'Pacific Div', 'Southwest Div', 
                'Pre All-Star', 'Post All-Star', 'Margin ≤3', 'Margin ≥10', 'Oct', 'Nov', 'Dec', 'Jan', 'Feb', 
                'Mar', 'Apr']

rank_20.head()

Rank Overall  Home   Road Eastern Conference  \
Team                                                                  
San Antonio Spurs         1   62-20  32-9  30-11               24-6   
Oklahoma City Thunder     2   59-23  34-7  25-16               23-7   
Los Angeles Clippers      3   57-25  34-7  23-18               21-9   
Indiana Pacers            4   56-26  35-6  21-20              38-14   
Houston Rockets           5   54-28  33-8  21-20               23-7   

                      Western Conference Atlantic Div Central Div  \
Team                                                                
San Antonio Spurs                  38-14          8-2         7-3   
Oklahoma City Thunder              36-16          8-2         8-2   
Los Angeles Clippers               36-16          9-1         7-3   
Indiana Pacers                     18-12         15-3        12-4   
Houston Rockets                    31-21          7-3         8-2   

                      Southeast Div Northwest Div ...  Post All-Star  \
Team                                              ...                  
San Antonio Spurs               9-1          11-7 ...           24-5   
Oklahoma City Thunder           7-3          11-5 ...          16-11   
Los Angeles Clippers            5-5          12-6 ...           20-7   
Indiana Pacers                 11-7           6-4 ...          16-14   
Houston Rockets                 8-2          11-7 ...          18-11   

                      Margin ≤3 Margin ≥10  Oct   Nov   Dec   Jan  Feb   Mar  \
Team                                                                           
San Antonio Spurs           7-0      35-12  1-0  13-3  11-4   8-6  9-3  16-0   
Oklahoma City Thunder      10-4       33-8  1-0  10-3  14-3  13-4  6-5  10-4   
Los Angeles Clippers        4-3       30-9  1-1  11-4   9-7  12-4  7-4  13-2   
Indiana Pacers              4-3      29-17  2-0  13-1  10-4  10-5  9-3  8-10   
Houston Rockets             7-6       30-9  1-0  12-5   8-8  10-4  8-2  10-4   

                       Apr  
Team                        
San Antonio Spurs      4-4  
Oklahoma City Thunder  5-4  
Los Angeles Clippers   4-3  
Indiana Pacers         4-3  
Houston Rockets        5-5  

[5 rows x 23 columns]

## Add features to the DataFrame

**Home Win**

In [52]:
# Determine whether the home team won
df_21['Home Win'] = df_21['Visitor Score'] < df_21['Home Score']

# Assign "class values" to the Home Win column
y_true = df_21['Home Win'].values

In [53]:
# Review new column with correctly populated fields
df.head()

Date Start Time (ET)          Visitor Team  Visitor Score  \
0 2013-10-29         7:00 pm         Orlando Magic             87   
1 2013-10-29        10:30 pm  Los Angeles Clippers            103   
2 2013-10-29         8:00 pm         Chicago Bulls             95   
3 2013-10-30         7:00 pm         Brooklyn Nets             94   
4 2013-10-30         8:30 pm         Atlanta Hawks            109   

             Home Team  Home Score  Box Score   OT  Attendance Notes  \
0       Indiana Pacers          97  Box Score  NaN       18165   NaN   
1   Los Angeles Lakers         116  Box Score  NaN       18997   NaN   
2           Miami Heat         107  Box Score  NaN       19964   NaN   
3  Cleveland Cavaliers          98  Box Score  NaN       20562   NaN   
4     Dallas Mavericks         118  Box Score  NaN       19834   NaN   

   Home Win  Home Last Win  Visitor Last Win  Home Win Streak  \
0      True          False             False                0   
1      True          False             False                0   
2      True          False             False                0   
3      True          False             False                0   
4      True          False             False                0   

   Visitor Win Streak  Home Team Ranks Higher  Home Team Won Last  
0                   0                    True                   0  
1                   0                   False                   0  
2                   0                    True                   0  
3                   0                   False                   0  
4                   0                   False                   0

**Home Last Win & Visitor Last Win**

In [54]:
# Create two new columns, with all of the values set to False
df_21['Home Last Win'] = False
df_21['Visitor Last Win'] = False

# Determining whether the home and visitor teams won their last games
# This will update the Home Last Win & Visitor Last Win columns to either remain False or change to True

from collections import defaultdict

won_last = defaultdict(int)   # The default value of int is 0

# Inefficient method to deterime previous wins
for index, row in df_21.iterrows():
    home_team = row['Home Team']
    visitor_team = row['Visitor Team']
    row['Home Last Win'] = won_last[home_team]
    row['Visitor Last Win'] = won_last[visitor_team]
    # Set current win
    won_last[home_team] = row['Home Win']
    won_last[visitor_team] = not row['Home Win']
df_21.sample(10)

Date Start Time (ET)           Visitor Team  Visitor Score  \
1043 2015-03-22         9:00 pm       Dallas Mavericks             92   
1123 2015-04-02         8:30 pm        Houston Rockets            108   
961  2015-03-12        10:30 pm        New York Knicks            101   
441  2014-12-26        10:00 pm           Phoenix Suns            115   
197  2014-11-24         7:00 pm   Los Angeles Clippers            113   
1098 2015-03-29        12:30 pm        Houston Rockets             99   
291  2014-12-06        10:00 pm          Orlando Magic            105   
48   2014-11-04         8:00 pm          Orlando Magic             90   
11   2014-10-29        10:30 pm  Oklahoma City Thunder             89   
87   2014-11-08         8:30 pm      Memphis Grizzlies             92   

                   Home Team  Home Score  Box Score   OT  Attendance Notes  \
1043            Phoenix Suns          98  Box Score  NaN     17435.0   NaN   
1123        Dallas Mavericks         101  Box Score  NaN     20062.0   NaN   
961       Los Angeles Lakers          94  Box Score  NaN     18997.0   NaN   
441         Sacramento Kings         106  Box Score  NaN     17317.0   NaN   
197        Charlotte Hornets          92  Box Score  NaN     17180.0   NaN   
1098      Washington Wizards          91  Box Score  NaN     20356.0   NaN   
291         Sacramento Kings          96  Box Score  NaN     16021.0   NaN   
48             Chicago Bulls          98  Box Score  NaN     21809.0   NaN   
11    Portland Trail Blazers         106  Box Score  NaN     19441.0   NaN   
87           Milwaukee Bucks          93  Box Score  NaN     13841.0   NaN   

      Home Win  Home Last Win  Visitor Last Win  
1043      True          False             False  
1123     False          False             False  
961      False          False             False  
441      False          False             False  
197      False          False             False  
1098     False          False             False  
291      False          False             False  
48        True          False             False  
11        True          False             False  
87        True          False             False

**Home Win Streak & Visitor Win Streak**

In [55]:
# Creating columns for winning streaks
df_21["Home Win Streak"] = 0
df_21["Visitor Win Streak"] = 0

# Did the home and visitor teams win their last game?

from collections import defaultdict

win_streak = defaultdict(int)

for index, row in df_21.iterrows():  # Note that this is not the most efficient method
    home_team = row["Home Team"]
    visitor_team = row["Visitor Team"]
    row["Home Win Streak"] = win_streak[home_team]
    row["Visitor Win Streak"] = win_streak[visitor_team]
    df_21.loc[index] = row    
    
    # Set current win streak
    if row["Home Win"]:
        win_streak[home_team] += 1
        win_streak[visitor_team] = 0
    else:
        win_streak[home_team] = 0
        win_streak[visitor_team] += 1
        
df_21.sample(10)

Date Start Time (ET)          Visitor Team  Visitor Score  \
840  2015-02-25         7:30 pm       New York Knicks             94   
112  2014-11-12        10:00 pm       Houston Rockets            113   
39   2014-11-02         6:00 pm       Toronto Raptors            102   
1202 2015-04-13         7:30 pm         Chicago Bulls            113   
832  2015-02-23         8:00 pm       Toronto Raptors             97   
1070 2015-03-25         7:00 pm        Indiana Pacers            103   
1215 2015-04-14        10:30 pm  Los Angeles Clippers            112   
1020 2015-03-20         7:30 pm        Indiana Pacers             92   
47   2014-11-03         7:00 pm       Houston Rockets            104   
436  2014-12-26         8:00 pm       Houston Rockets            117   

                   Home Team  Home Score  Box Score   OT  Attendance  \
840           Boston Celtics         115  Box Score  NaN     16899.0   
112   Minnesota Timberwolves         101  Box Score  NaN     18996.0   
39                Miami Heat         107  Box Score  NaN     19666.0   
1202           Brooklyn Nets          86  Box Score  NaN     17732.0   
832     New Orleans Pelicans         100  Box Score  NaN     16514.0   
1070      Washington Wizards         101  Box Score  NaN     18514.0   
1215            Phoenix Suns         101  Box Score  NaN     18055.0   
1020     Cleveland Cavaliers          95  Box Score  NaN     20562.0   
47        Philadelphia 76ers          93  Box Score  NaN     12896.0   
436        Memphis Grizzlies         111  Box Score   OT     18119.0   

                      Notes  Home Win  Home Last Win  Visitor Last Win  \
840                     NaN      True          False             False   
112   at Mexico City Mexico     False          False             False   
39                      NaN      True          False             False   
1202                    NaN     False          False             False   
832                     NaN      True          False             False   
1070                    NaN     False          False             False   
1215                    NaN     False          False             False   
1020                    NaN      True          False             False   
47                      NaN     False          False             False   
436                     NaN     False          False             False   

      Home Win Streak  Visitor Win Streak  
840                 1                   0  
112                 0                   0  
39                  2                   2  
1202                0                   2  
832                 1                   0  
1070                0                   0  
1215                0                   6  
1020                1                   0  
47                  0                   3  
436                 0                   1

______________________________________________________________________________________________________________________

**Home Team Ranks Higher**

In [56]:
# Build function to add Home Team Ranks Higher feature
def home_team_ranks_higher_2015(row):
    home_team = row['Home Team']
    visitor_team = row['Visitor Team']
    
    # Adjusting the Charlotte team names due to off-season league changes between 13-14 & 14-15
    if home_team == "Charlotte Hornets":
        home_team = "Charlotte Bobcats"
    if visitor_team == "Charlotte Hornets":
        visitor_team = "Charlotte Bobcats"
    
    home_rank = rank_20.loc[home_team]['Rank']
    visitor_rank = rank_20.loc[visitor_team]['Rank']
    
    # The higher ranking will be the lower number
    return home_rank < visitor_rank

In [57]:
# Create new column by using .apply(home_team_ranks_higher_2015) across the DataFrame
df_21["Home Team Ranks Higher"] = df_21.apply(home_team_ranks_higher_2015, axis = 1)
df_21.head()

Date Start Time (ET)      Visitor Team  Visitor Score  \
0 2014-10-28        10:30 pm   Houston Rockets            108   
1 2014-10-28         8:00 pm     Orlando Magic             84   
2 2014-10-28         8:00 pm  Dallas Mavericks            100   
3 2014-10-29         7:30 pm     Brooklyn Nets            105   
4 2014-10-29         7:00 pm   Milwaukee Bucks            106   

              Home Team  Home Score  Box Score   OT  Attendance Notes  \
0    Los Angeles Lakers          90  Box Score  NaN     18997.0   NaN   
1  New Orleans Pelicans         101  Box Score  NaN     17097.0   NaN   
2     San Antonio Spurs         101  Box Score  NaN     19615.0   NaN   
3        Boston Celtics         121  Box Score  NaN     18624.0   NaN   
4     Charlotte Hornets         108  Box Score   OT     19439.0   NaN   

   Home Win  Home Last Win  Visitor Last Win  Home Win Streak  \
0     False          False             False                0   
1      True          False             False                0   
2      True          False             False                0   
3      True          False             False                0   
4      True          False             False                0   

   Visitor Win Streak  Home Team Ranks Higher  
0                   0                   False  
1                   0                    True  
2                   0                    True  
3                   0                   False  
4                   0                    True

In [58]:
# Add new column without using a function & .apply
# df_21["Home Team Ranks Higher"] = 0

# for index, row in df_21.iterrows():
#     home_team = row["Home Team"]
#     visitor_team = row["Visitor Team"]
   
    # Adjusting the New Orleans team names due to off-season league changes between 12-13 & 13-14
#     if home_team == "Charlotte Hornets":
#         home_team = "Charlotte Bobcats"
#     elif visitor_team == "Charlotte Hornets":
#         visitor_team = "Charlotte Bobcats"
    
#     home_rank = rank_20[rank_20["Team"] == home_team]["Rank"].values[0]
#     visitor_rank = rank_20[rank_20["Team"] == visitor_team]["Rank"].values[0]
#     row["Home Team Ranks Higher"] = int(home_rank > visitor_rank)
#     df_21.loc[index] = row

______________________________________________________________________________________________________________________

**Home Team Won Last**

In [59]:
# Add a new feature without using function -> Determines whether a team won the last matchup between the 2 teams
# This does not take into consideration the home/visitor teams

last_match_winner = defaultdict(int)
df_21["Home Team Won Last"] = 0

for index, row in df_21.iterrows():
    home_team = row["Home Team"]
    visitor_team = row["Visitor Team"]
    teams = tuple(sorted([home_team, visitor_team]))  # Sort for a consistent ordering
    
    # Set in the row, who won the last encounter
    row["Home Team Won Last"] = 1 if last_match_winner[teams] == row["Home Team"] else 0
    df_21.loc[index] = row
    
    # Who won this one?
    winner = row["Home Team"] if row["Home Win"] else row["Visitor Team"]
    last_match_winner[teams] = winner
    
df_21.sample(7)

Date Start Time (ET)           Visitor Team  Visitor Score  \
527  2015-01-07         9:00 pm          Orlando Magic             90   
1157 2015-04-07         8:00 pm      San Antonio Spurs            113   
857  2015-02-27         7:30 pm        New York Knicks            121   
701  2015-01-30         9:00 pm  Golden State Warriors            100   
1085 2015-03-28         8:00 pm        New York Knicks             80   
1223 2015-04-15         9:30 pm         Indiana Pacers             83   
52   2014-11-04         8:00 pm      Charlotte Hornets             91   

                  Home Team  Home Score  Box Score   OT  Attendance Notes  \
527          Denver Nuggets          93  Box Score  NaN     13513.0   NaN   
1157  Oklahoma City Thunder          88  Box Score  NaN     18203.0   NaN   
857         Detroit Pistons         115  Box Score  2OT     16182.0   NaN   
701               Utah Jazz         110  Box Score  NaN     19295.0   NaN   
1085          Chicago Bulls         111  Box Score  NaN     22152.0   NaN   
1223      Memphis Grizzlies          95  Box Score  NaN     17109.0   NaN   
52     New Orleans Pelicans         100  Box Score  NaN     14840.0   NaN   

      Home Win  Home Last Win  Visitor Last Win  Home Win Streak  \
527       True          False             False                2   
1157     False          False             False                0   
857      False          False             False                0   
701       True          False             False                0   
1085      True          False             False                2   
1223      True          False             False                0   
52        True          False             False                0   

      Visitor Win Streak  Home Team Ranks Higher  Home Team Won Last  
527                    0                    True                   0  
1157                   7                   False                   0  
857                    0                   False                   1  
701                    0                   False                   0  
1085                   0                    True                   1  
1223                   6                   False                   1  
52                     0                   False                   0

In [60]:
# Review the added features in the 2014-2015 season DataFrame
df_21[['Date', 'Visitor Team', 'Visitor Score', 'Home Team', 'Home Score', 'Home Win', 'Home Last Win', 
       'Visitor Last Win', 'Home Team Ranks Higher', 'Home Win Streak', 'Visitor Win Streak', 
       'Home Team Won Last']][:5]

Date      Visitor Team  Visitor Score             Home Team  \
0 2014-10-28   Houston Rockets            108    Los Angeles Lakers   
1 2014-10-28     Orlando Magic             84  New Orleans Pelicans   
2 2014-10-28  Dallas Mavericks            100     San Antonio Spurs   
3 2014-10-29     Brooklyn Nets            105        Boston Celtics   
4 2014-10-29   Milwaukee Bucks            106     Charlotte Hornets   

   Home Score  Home Win  Home Last Win  Visitor Last Win  \
0          90     False          False             False   
1         101      True          False             False   
2         101      True          False             False   
3         121      True          False             False   
4         108      True          False             False   

   Home Team Ranks Higher  Home Win Streak  Visitor Win Streak  \
0                   False                0                   0   
1                    True                0                   0   
2                    True                0                   0   
3                   False                0                   0   
4                    True                0                   0   

   Home Team Won Last  
0                   0  
1                   0  
2                   0  
3                   0  
4                   0

In [61]:
# Use selected features as input for the classifier (target variable) for if the home team is ranked higher
X_home_higher_15 = df_21[['Home Last Win', 'Visitor Last Win', 'Home Team Ranks Higher', 'Home Team Won Last']].values

______________________________________________________________________________________________________________________

***There was a team name change in the test set of data, causing the previous encoder to not carry over. HOW TO FIX?***

***Start at 30:58 in lecture***

Options:
1. Re-clean the data so that the team names are the same. This will create an inaccuracy in seasonal team names, but keep the encoded data formatted as is
2. Pass an explicit mapping to ordinal encoding and encode the two versions of each name as the same category. Then  pass the now a encoded integer column into any other method. This would keep the integrity of the data across the seasons, but still allow for further analysis

In [ ]:
# Map the previously encoded standings for the Charlotte Hornets (new) to be assigned to the same encoding as 
# the Charlotte Bobcats (old)


In [ ]:
# Reuse the trained team encoding and onehot classifier
# encoding = LabelEncoder()
# encoding.fit(df['Home Team].values)

home_teams_15 = encoding.transform(df_21['Home Team'].values)
visitor_teams_15 = encoding.transform(df_21['Visitor Team'].values)

X_teams_15 = np.vstack([home_teams_15, visitor_teams_15]).T
X_teams_15 = onehot.transform(X_teams_15).todense()

In [ ]:
X_all_15 = np.hstack([X_home_higher_15, X_teams_15])
X_all_15.shape

## Setting the Baseline

In [ ]:
print('The home team wins {:.2f}% of the games'.format(100 * np.mean(y_true)))

y_pred = [1] * len(y_test)

print('F1: {:.4f}'.format(f1_score(y_test, y_pred, pos_label = None, average = 'weighted')))
print(classification_report(y_test, y_pred))

In [ ]:
grid.fit(X_all, y_true)

y_pred = grid.predict(X_all_15)

In [1]:
from sklearn import classification_report

print(classification_report(y_test, y_pred))

ImportError: cannot import name 'classification_report' from 'sklearn' (C:\Users\thoma\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\sklearn\__init__.py)

In [ ]:
print('This results in getting {:.1f}% of predictions correct!'.format(100 * np.mean(y_pred == y_test)))

______________________________________________________________________________________________________________________